pkg/reconciler/apiresource/controller.go

```diff
@@ -0,0 +1,373 @@
+limitations under the License.
+*/
+
+package apiresource
+
+import (
+	"context"
+	"io/ioutil"
+	"time"
+
+	apiresourcev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	typedapiresource "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/typed/apiresource/v1alpha1"
+	kcpexternalversions "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	apiresourcelister "github.com/kcp-dev/kcp/pkg/client/listers/apiresource/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/util/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	typedapiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	crdexternalversions "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions"
+	crdlister "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8serorrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"sigs.k8s.io/yaml"
+)
+
+const resyncPeriod = 10 * time.Hour
+const ClusterNameAndGVRIndexName = "clusterNameAndGVR"
+
+func GetClusterNameAndGVRIndexKey(clusterName string, gvr metav1.GroupVersionResource) string {
+	return clusterName + "$" + gvr.String()
+}
+
+func NewController(cfg *rest.Config, autoPublishNegotiatedAPIResource bool) *Controller {
+	apiresourceClient := typedapiresource.NewForConfigOrDie(cfg)
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	stopCh := make(chan struct{}) // TODO: hook this up to SIGTERM/SIGINT
+
+	crdClient := typedapiextensions.NewForConfigOrDie(cfg)
+
+	c := &Controller{
+		queue:                            queue,
+		apiresourceClient:                apiresourceClient,
+		crdClient:                        crdClient,
+		stopCh:                           stopCh,
+		AutoPublishNegotiatedAPIResource: autoPublishNegotiatedAPIResource,
+	}
+
+	apiresourceSif := kcpexternalversions.NewSharedInformerFactoryWithOptions(kcpclient.NewForConfigOrDie(cfg), resyncPeriod)
+	apiresourceSif.Apiresource().V1alpha1().NegotiatedAPIResources().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(Add, nil, obj) },
+		UpdateFunc: func(oldObj, obj interface{}) { c.enqueue(Update, oldObj, obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueue(Delete, nil, obj) },
+	})
+	c.negotiatedApiResourceIndexer = apiresourceSif.Apiresource().V1alpha1().NegotiatedAPIResources().Informer().GetIndexer()
+	c.negotiatedApiResourceIndexer.AddIndexers(map[string]cache.IndexFunc{
```

Comment:
Do we need to define our own indexer here and below?

Reply:
Well, we could possibly do without it, but it makes it more efficient, and also easier to code, to setup indexer instead of listing all instances and filtering them out all over the place.